[AMDGPU][CodeGen] Improve handling of memcpy for -Os/-Oz compilations

shiltian opened the pull request:
We had some instances when LLVM would not inline fixed-count memcpy and ended up
attempting to lower it a a libcall, which would not work on AMDGPU as the
address space doesn't meet the requirement, causing compiler crash.

The patch relaxes the threshold used for -Os/-Oz compilation so we're always allowed
to inline memory copy functions.

This patch basically does the same thing as https://reviews.llvm.org/D158226 for
AMDGPU.

Fix #88497.

arsenm commented on llvm/test/CodeGen/AMDGPU/memcpy-libcall.ll:
```diff
@@ -0,0 +1,642 @@
+
+declare void @llvm.memcpy.p0.p0.i64(ptr noalias nocapture writeonly, ptr noalias nocapture readonly, i64, i1 immarg) #0
+
+define dso_local amdgpu_kernel void @_Z11copy_globalPvS_(ptr addrspace(1) nocapture noundef writeonly %dest.coerce, ptr addrspace(1) nocapture noundef readonly %src.coerce) local_unnamed_addr #0 {
```
Drop all the dso_local/local_unnamed_addr

arsenm commented on llvm/test/CodeGen/AMDGPU/memcpy-libcall.ll:
```diff
@@ -0,0 +1,642 @@
+
+declare void @llvm.memcpy.p0.p3.i64(ptr noalias nocapture writeonly, ptr addrspace(3) noalias nocapture readonly, i64, i1 immarg)
+
+attributes #0 = { minsize }
```
Should you also test optsize?

arsenm reviewed (commented):
Should we change the expansion threshold in the PreISelIntrinsicLowering?

arsenm commented on llvm/test/CodeGen/AMDGPU/memcpy-libcall.ll:
```diff
@@ -0,0 +1,642 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
+; RUN: llc -march=amdgcn -mcpu=gfx908 %s -o - | FileCheck %s
```
use -mtriple=amdgcn-amd-amdhsa so we don't get the weird scratch ABI

shiltian commented:
> Should we change the expansion threshold in the PreISelIntrinsicLowering?

Not really in this patch. With this patch those intrinsics will be lowered before instruction selection.

shiltian commented:
gentle ping

shiltian commented:
gentle ping + 1

Pierre-vh commented on llvm/test/CodeGen/AMDGPU/memcpy-libcall.ll:
```diff
@@ -0,0 +1,973 @@
+@shared = addrspace(3) global %struct.S undef, align 4
+
+; Function Attrs: minsize
+define void @_Z12copy_genericPvPKv__minsize(ptr nocapture noundef writeonly %dest, ptr nocapture noundef readonly %src) #0 {
```
Rename functions to something generic (can use metarenamer pass); some attributes (both parameters and function) can probably be simplified or removed as well?

Pierre-vh commented on llvm/test/CodeGen/AMDGPU/memcpy-libcall.ll:
```diff
@@ -0,0 +1,973 @@
+; Function Attrs: minsize
+define void @_Z12copy_genericPvPKv__minsize(ptr nocapture noundef writeonly %dest, ptr nocapture noundef readonly %src) #0 {
+; CHECK-LABEL: _Z12copy_genericPvPKv__minsize:
+; CHECK:       ; %bb.0: ; %entry
```
Can you precommit this test so we can see before/after?

shiltian replied:
before: crash Lol

Pierre-vh replied:
Oops, I should have read the commit message first, my bad
![](https://cdn3.emoji.gg/emojis/8637-sadcat-ok.png)

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp:
```diff
@@ -59,6 +59,12 @@ unsigned AMDGPUTargetLowering::numBitsSigned(SDValue Op, SelectionDAG &DAG) {
     : TargetLowering(TM), Subtarget(&STI) {
+  // Always lower memset, memcpy, and memmove intrinsics to load/store
+  // instructions, rather then generating calls to memset, mempcy or memmove.
+  MaxStoresPerMemset = MaxStoresPerMemsetOptSize = ~0U;
```
Can you document where the `~0U` comes from?
Is it a magic value recognized in this pass or somewhere else?

Could we eventually have a constant variable for that value so its source is clear?

arsenm replied:
If you wanted to be fancy you could use `std::numeric_limits<typeof(MaxStoresPerMemset)>::max()`

Pierre-vh replied:
> This is not a magic number, it is basically inf. These numbers are used as a threshold that determines whether the three functions will be lowered to load/store or lib call. Since we don't have lib call anyway, we need to lower them in all cases.

I see, `~0U` works for me then. Maybe just add a small comment on top explaining that this is the libcall lowering threshold and it shouldn't be reachable?

ex-rzr commented:
Is this PR related to https://github.com/llvm/llvm-project/issues/88497?

ex-rzr commented:
Yes, it fixes my issue, see https://github.com/llvm/llvm-project/issues/88497#issuecomment-2058732364

arsenm commented on llvm/test/CodeGen/AMDGPU/memcpy-libcall.ll:
```diff
@@ -0,0 +1,2717 @@
+  ret void
+}
+
+; Function Attrs: nocallback nofree nounwind willreturn memory(argmem: readwrite)
```
Drop attribute comments

arsenm commented:
Can you add the fixes issue number to the description

shiltian committed:
[AMDGPU][CodeGen] Improve handling of memcpy for -Os/-Oz compilations

We had some instances when LLVM would not inline fixed-count memcpy and ended up
attempting to lower it a a libcall, which would not work on AMDGPU as the
address space doesn't meet the requirement, causing compiler crash.

The patch relaxes the threshold used for -Os compilation so we're always allowed
to inline memory copy functions.

This patch basically does the same thing as https://reviews.llvm.org/D158226 for
AMDGPU.
